packages/@react-stately/data/src/useAsyncList.ts

```diff
@@ -253,17 +273,36 @@ export function useAsyncList<T, C = string>(options: AsyncListOptions<T, C>): As
     let abortController = new AbortController();
     try {
       dispatch({...action, abortController});
+      let previousFilterText = action.filterText ?? data.filterText;
+
+      if (action.type !== 'success' && action.type !== 'update') {
+        loadingState.current = action.type;
+      } else {
+        loadingState.current = 'idle';
+      }
+
       let response = await fn({
         items: data.items.slice(),
         selectedKeys: data.selectedKeys,
         sortDescriptor: action.sortDescriptor ?? data.sortDescriptor,
         signal: abortController.signal,
         cursor: action.type === 'loadingMore' ? data.cursor : null,
-        filterText: action.filterText ?? data.filterText
+        filterText: previousFilterText,
+        loadingState: loadingState.current
       });
 
+      let filterText = response.filterText ?? previousFilterText;
+      loadingState.current = 'idle';
       dispatch({type: 'success', ...response, abortController});
+
+      // Fetch a new filtered list if filterText is updated via `load` response func rather than list.setFilterText
+      // Only do this if not aborted (e.g. user triggers another filter action before load completes)
+      if (filterText && (filterText !== previousFilterText) && !abortController.signal.aborted) {
+        loadingState.current = 'filtering';
```

Comment:
Should be set inside `dispatchFetch` already right?

Reply:
bleh, was too broad with my search on "dispatch", good catch